keps/sig-cli/4764-kubectl-watchlist/README.md

```diff
@@ -0,0 +1,757 @@
+Pick one more of these and delete the rest.
+Please describe all items visible to end users below with sufficient detail so that they can verify correct enablement
+and operation of this feature.
+Recall that end users cannot usually observe component logs or access metrics.
+-->
+
+- [ ] Events
+  - Event Reason: 
+- [ ] API .status
+  - Condition name: 
+  - Other field: 
+- [ ] Other (treat as last resort)
+  - Details:
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+<!--
+This is your opportunity to define what "normal" quality of service looks like
+for a feature.
+
+It's impossible to provide comprehensive guidance, but at the very
+high level (needs more precise definitions) those may be things like:
+  - per-day percentage of API calls finishing with 5XX errors <= 1%
+  - 99% percentile over day of absolute value from (job creation time minus expected
+    job creation time) for cron job <= 10%
+  - 99.9% of /health requests per day finish with 200 code
+
+These goals will help you determine what you need to measure (SLIs) in the next
+question.
+-->
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+<!--
+Pick one more of these and delete the rest.
+-->
+
+- [ ] Metrics
+  - Metric name:
+  - [Optional] Aggregation method:
+  - Components exposing the metric:
+- [ ] Other (treat as last resort)
+  - Details:
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+<!--
+Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
+implementation difficulties, etc.).
+-->
+
+### Dependencies
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### Does this feature depend on any specific services running in the cluster?
+
+<!--
+Think about both cluster-level services (e.g. metrics-server) as well
+as node-level agents (e.g. specific version of CRI). Focus on external or
+optional services that are needed. For example, if this feature depends on
+a cloud provider API, or upon an external software-defined storage or network
+control plane.
+
+For each of these, fill in the following—thinking about running existing user workloads
+and creating new ones, as well as about cluster-level services (e.g. DNS):
+  - [Dependency name]
+    - Usage description:
+      - Impact of its outage on the feature:
+      - Impact of its degraded performance or high-error rates on the feature:
+-->
+
+### Scalability
+
+<!--
+For alpha, this section is encouraged: reviewers should consider these questions
+and attempt to answer them.
+
+For beta, this section is required: reviewers must answer these questions.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+-->
+
+###### Will enabling / using this feature result in any new API calls?
+
```

Comment:
:thought_balloon: Is there a way for the denial of the **watch** to include a hint that a **list** would also fail;
for example, this hint could be included whenever `sendInitialEvents` is set and the API server is able to dry-run
check the equivalent **list** verb.

Reply:
I don't quite understand what you mean by hint? 

When **watchlist** failed, we can give a warning message and then fallback to **list**.  And When rolling back to **list**, the behavior of kubectl should be consistent with the existing behavior, that is, the results of **list** are returned. I'm not sure about what this hint used to tell user?

Reply:
oh I see the point now. you mean when the **watchlist** fails, the error is returned with a hint, to tell the client **list** would also fail. so it's unnecessary for client to try **list** again. Did I understand correctly?

I don't believe that we have a easy solution for this. As in my understand, this case can only happen when user have no permission to **list** resource. however, URL query parameters are decoded in the `Handler`, but authz happens before that, in the `Filters`. 

A simpler and feasible solution is to initiate a `SubjectAccessReview` during **watchlist** and return the result as Hint. However, this will bring additional consumption to all **watchlist**, and it seems to be not much different from the client re-initiating the **list** request. After all, if the failure is caused by authz, the **list** has not actually started processing at this time, and it will not consume too many resources.

Reply:
or we can consider the **watch** permission as a superset of the **list** permission. After all, users with the **watch** permission actually already have the **list** permission. In this case, as long as the user can pass the **watch** authz, they will be able to perform subsequent **list** operations for sure. 

However, this is an incompatible change. I am not sure it is appropriate to discuss it in this KEP.

Reply:
Kind of. `WatchList` is a client method not a logical API verb, though (so `WatchList` calls either **watch** or **list**).

Reply:
> initiate a SubjectAccessReview during watchlist and return the result as Hint

I was thinking of this, perhaps with a fast timeout (eg end the attempt after 10ms, and in that case don't include a hint)

Reply:
I agree with that the semantics is clearer indeed, if there is only one API call. And it's make sense to initiate a `SubjectAccessReview` inside the API server for saving the time of an RPC call between api and client. However fallback should be rara by design, but this machanism will still slow down every **watch** request(even with fast timeout). Although the impact could be very little, is it worth it?

In addition, this seems to be more of an enhancement to `WatchList`, rather than specific to kubectl? If we need to add this optimization, then the `reflector` in `client-go` should also have the same mechanism? I'm not sure if this should be put in `KEP-3157`?

Reply:
Let's resolve this. I used the :thought_balloon: emoji in https://github.com/kubernetes/enhancements/pull/4767#discussion_r1693983129 to try to imply that this was an aside, rather than any kind of blocking feedback.

Reply:
oh, I see. thanks for your explanation. 😊 

what about if API server initiate a `SubjectAccessReview` and returns a hint only if `WatchList` is disabled or **watch** permission deny? this could minimize the extra authz query.

Reply:
Yes, that's how I'd imagined it.

Reply:
What we discussed earlier have been updated into the KEP. PTAL